controller/lib/hal/stepper.h

```diff
@@ -0,0 +1,342 @@
+  MOVE_NEG = 0x40,     // Move - steps, steps is given as a parameter
+  MOVE_POS = 0x41,     // Move + steps, steps is given as a parameter
+  GOTO = 0x60,         // Go to an absolute position using shortest path
+  GOTO_NEG = 0x68,     // Go to an absolute position in negative direction
+  GOTO_POS = 0x69,     // Go to an absolute position in positive direction
+  HOME = 0x70,         // Same as GOTO 0
+  RESET_POS = 0xD8,    // Set absolute position register to 0
+  RESET_DEVICE = 0xC0, // Reset the stepper chip
+  SOFT_STOP = 0xB0,    // Perform a soft stop
+  HARD_STOP = 0xB8,    // Perform a hard stop
+  SOFT_DISABLE = 0xA0, // Soft stop, then disable
+  HARD_DISABLE = 0xA8, // Hard stop, then disable
+  GET_STATUS = 0xD0,   // Read the 16-bit status word
+};
+
+enum class StepMtrParam {
+  ABS_POS = 0x01,           // Absolute position
+  ELE_POS = 0x02,           // Electrical position
+  MARK_POS = 0x03,          // Mark position
+  SPEED = 0x04,             // Current speed
+  ACCEL = 0x05,             // Acceleration
+  DECEL = 0x06,             // Deceleration
+  MAX_SPEED = 0x07,         // Maximum allowed speed
+  MIN_SPEED = 0x08,         // Minimum allowed speed
+  KVAL_HOLD = 0x09,         //
+  KVAL_RUN = 0x0A,          //
+  KVAL_ACCEL = 0x0B,        //
+  KVAL_DECEL = 0x0C,        //
+  INT_SPEED = 0x0D,         //
+  START_SLOPE = 0x0E,       //
+  FINAL_SLOPE_ACCEL = 0x0F, //
+  FINAL_SLOPE_DECEL = 0x10, //
+  THERMAL_COMP = 0x11,      //
+  ADC_OUT = 0x12,           //
+  OVER_CRNT_THRESH = 0x13,  //
+  STALL_THRESH = 0x14,      //
+  FULL_STEP_SPEED = 0x15,   // Speed at which to switch to full step mode
+  STEP_MODE = 0x16,         //
+  ALARM_ENA = 0x17,         //
+  GATE_CFG1 = 0x18,         //
+  GATE_CFG2 = 0x19,         //
+  CONFIG = 0x1A,            //
+  STATUS = 0x1B,            //
+};
+
+// Error codes returned by my functions
+enum class StepMtrErr {
+  OK,
+  BAD_PARAM,     // The parameter ID is invalid
+  BAD_VALUE,     // Illegal value passed
+  WOULD_BLOCK,   // Call would block, can't be called from within ISR
+  QUEUE_FULL,    // Can't add command to queue, not enough space
+  INVALID_STATE, // Invalid state for command.
+};
+
+// States for the stepper motor communication interface
+enum class StepCommState {
+  IDLE,        // Not currently communicating
+  SEND_QUEUED, // Sending data that was queued up by control loop
+  SEND_SYNC,   // Sending data that the background thread is waiting on.
+};
+
+enum class StepMoveStatus {
+  STOPPED,
+  ACCELERATING,
+  DECELERATING,
+  CONSTANT_SPEED,
+};
+
+// Detailed status about the stepper driver chip.
+// Note, fields marked with * below are latched meaning that they
+// will return true if the event has occurred since the last time
+// the status was read.  The act of reading the status clears them
+// so they won't be true on the next read unless they happen again.
+struct StepperStatus {
+
```

Comment:
I'll add a constructor although I'm not sure it's meaningful.  The only use of the status is to read it from this driver which initializes all the fields.

Reply:
No need for constructor.
Default values for any struct or class can just be like this
```
int member_variable_ {0};
```
and it will by default be initialized to that.
Having default values can avoid unpredictable behavior. It may never happen if the struct is used as intended, but we are insuring against those other unintentional situations.

Reply:
I'm afraid I mostly code in C in my day job and haven't been keeping up to date on the changes to the C++ language.  I didn't realize I could initialize member variables like that.
Certainly much nicer that way.